plasmapy/atomic/particle_input.py

```diff
@@ -0,0 +1,169 @@
+    ----------
+
+    wrapped_function : callable
+        The function to be decorated.
+
+    must_be : str, set, list, or tuple; optional
+        The categories that the Particle must be in; otherwise the
+
+
+    Notes
+    -----
+
+    This version of particle_input only works with functions, but in the future
+    should be extended to work with classes and methods.
+
+    """
+
+    def _errmsg(argname: str, argval: str, Z: int = None,
+                mass_numb: int = None, funcname: str = None) -> str:
+        r"""Returns a string with an appropriate error message for an
+        InvalidParticleError."""
+
+        errmsg = f"In {funcname}, {argname} = {repr(argval)} "
+        if mass_numb is not None or Z is not None:
+            errmsg += "with "
+        if mass_numb is not None:
+            errmsg += f"mass_numb = {repr(mass_numb)} "
+        if mass_numb is not None and Z is not None:
+            errmsg += "and "
+        if Z is not None:
+            errmsg += f"integer charge Z = {repr(Z)} "
+        errmsg += "does not correspond to a valid particle."
+        return errmsg
+
+    def decorator(wrapped_function: Callable):
+        wrapped_signature = inspect.signature(wrapped_function)
+
+        @functools.wraps(wrapped_function)
+        def wrapper(*args, **kwargs):
+
+            # The following couple of statements will likely need to be
+            # modified in order to work with methods.
+
+            bound_args = wrapped_signature.bind(*args, **kwargs)
+            arguments = bound_args.arguments
+            argnames = bound_args.arguments.keys()
+            funcname = wrapped_function.__name__
+            annotations = wrapped_function.__annotations__
+
+            args_to_become_particles = [
+                argname for argname in annotations.keys()
+                if annotations[argname] is Particle
+            ]
+
+            if not args_to_become_particles:
+                raise AtomicError(
+                    f"None of the arguments or keywords to {funcname} "
+                    f"have been annotated with Particle as required "
+                    f"by the particle_input decorator.")
+            elif len(args_to_become_particles) > 1:
+                if 'Z' in argnames or 'mass_numb' in argnames:
+                    raise AtomicError(
+                        f"The arguments Z and mass_numb in {funcname} are not "
+                        f"allowed when more than one argument or keyword is "
+                        f"annotated with Particle in functions decorated "
+                        f"with @particle_input.")
+
+            # If the number of arguments and keywords annotated with
+            # Particle is exactly one, then the Z and mass_numb keywords
+            # can be used without potential for ambiguity.
+
+            Z = arguments.get('Z', None)
+            mass_numb = arguments.get('mass_numb', None)
+
+            if Z is not None and not isinstance(Z, int):
```

Comment:
Do you need to do this here? Won't the constructor of the `Particle` class do it?

Reply:
I see why you're taking exception to this exception.  The constructor of the `Particle` class does take care of all of this, so I can delete this code.